gtsfm/evaluation/metrics.py

```diff
@@ -0,0 +1,362 @@
+        return summary
+
+    def _get_distribution_quartiles(self, data: np.ndarray) -> Dict[int, float]:
+        """Computes quartiles for the provided 1D data distribution.
+
+        Args:
+            data: 1D distribution of metric values
+
+        Returns:
+            Quartiles of the data as a dict where keys are q0, q1, q2, q3, and q4
+        """
+        query = list(range(0, 101, 25))
+        quartiles = np.percentile(data, query)
+        output = {}
+        for i, q in enumerate(query):
+            output["q" + str(i)] = quartiles[i].tolist()
+        return output
+
+    def get_metric_as_dict(self) -> Dict[str, Any]:
+        """Provides a dict based representation of the metric that can be serialized to JSON.
+
+        The dict contains a single element, for which the key is the name of the metric.
+        If metric is a distribution, the dict is in the below format: 
+        {
+            metric_name: {
+               DATA_KEY: {
+                   .. raw data if stored ..
+               }
+               SUMMARY_KEY: {
+                    .. summary (stats) of distribution ..    
+               }
+            }
+        }
+        If the metric is scalar, it is stored simply as {metric_name: value}. 
+
+        Returns:
+            The metric as a dict representation explained above.
+        """
+        if self._dim == 0:
+            return {self._name: self._data.tolist()}
+
+        metric_dict = {SUMMARY_KEY: self.summary}
+        if self._data is not None:
+            metric_dict[DATA_KEY] = self._data.tolist()
+        return {self._name: metric_dict}
+
+    def save_to_json(self, json_filename: str) -> None:
+        """Saves this metric's dict representation to a JSON file.
+
+        Args:
+            Path to the json file.
+        """
+        io.save_json_file(json_filename, self.get_metric_as_dict())
+
+    @classmethod
+    def parse_from_dict(cls, metric_dict: Dict[str, Any]) -> GtsfmMetric:
+        """Creates a GtsfmMetric by parsing a dict representation.
+
+        It is assumed that the dict representation is the format created by GtsfmMetric.
+
+        Args:
+            metric_dict: Dict representation of the metric.
+
+        Returns:
+            Parsed GtsfmMetric instance.
+        """
+        if len(metric_dict) != 1:
+            raise AttributeError("Input metric dict should have a single key-value pair.")
+
+        metric_name = list(metric_dict.keys())[0]
+        metric_value = metric_dict[metric_name]
+
+        # 1D distribution metrics
+        if isinstance(metric_value, dict):
+            data = None
+            summary = None
+            if DATA_KEY in metric_value:
+                data = metric_value[DATA_KEY]
+            if SUMMARY_KEY in metric_value:
+                summary = metric_value[SUMMARY_KEY]
+            return cls(metric_name, data=data, summary=summary)
+
+        # Scalar metrics
+        return cls(metric_name, metric_value)
+
+
```

Comment:
this can be a named tuple?

Reply:
it would need to be a `dataclass` not `NamedTuple` if we want it to be mutable, so we could call `.extend()`

Reply:
i think it's fine currently, though